source/dynamical_systems/include/dynamical_systems/IDynamicalSystem.hpp

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <list>
+#include <map>
+#include <memory>
+
+#include "state_representation/space/cartesian/CartesianPose.hpp"
+#include "state_representation/space/cartesian/CartesianState.hpp"
+#include "state_representation/parameters/ParameterInterface.hpp"
+
+/**
+ * @namespace dynamical_systems
+ * @brief Systems of equations relating state variables to their derivatives
+ */
+namespace dynamical_systems {
+
+/**
+ * @class IDynamicalSystem
+ * @brief Abstract class for a dynamical system.
+ * @tparam S Underlying state type of the dynamical system
+ */
+template<class S>
+class IDynamicalSystem {
+public:
+  /**
+   * @brief Empty constructor
+   */
+  IDynamicalSystem();
+
+  /**
+   * @brief Check compatibility between a state and the dynamical system.
+   * @param state The state to check for compatibility
+   * @return True if the state is compatible with the dynamical system
+   */
+  [[nodiscard]] virtual bool is_compatible(const S& state) const;
+
+  /**
+   * @brief Evaluate the value of the dynamical system at a given state.
+   * @param state State at which to perform the evaluation
+   * @return The resulting state (velocity) of the dynamical system
+   */
+  [[nodiscard]] S evaluate(const S& state) const;
+
+  /**
+   * @brief Return the base frame of the dynamical system.
+   * @return The base frame
+   */
+  [[nodiscard]] S get_base_frame() const;
+
+  /**
+   * @brief Set the base frame of the dynamical system.
+   * @param base_frame The new base frame
+   */
+  virtual void set_base_frame(const S& base_frame);
+
+  /**
+   * @brief Get a parameter of the dynamical system by its name.
+   * @tparam T Type of the parameter
+   * @param name The name of the parameter
+   * @return The parameter, if it exists
+   */
+  template<typename T>
+  T get_parameter(const std::string& name);
+
+  /**
+   * @brief Get a list of all the parameters of the dynamical system.
+   * @return The list of parameters
+   */
+  [[nodiscard]] std::list<std::shared_ptr<state_representation::ParameterInterface>> get_parameters() const;
```

Comment:
In `modulo` parameters are actually stored as a `map` where the key is the name or the parameter. In fact this could return a `map` as well but the reason I opted for a list here is that you just need to go trough all the parameters returned by the class and store them in a `map` at the `Cell` level. The is a function `add_parameters` in `Cell` that takes this list, declare them in `ros` and store them in the `map`. Obviously this function could also take a map if that is preferable,

Reply:
Then we should just return the map here, that will be easier for everyone. It will already be a breaking change anyway if this is released as 5.0